modules/training_pipeline/training_pipeline/prompt_templates/prompter.py

```diff
@@ -0,0 +1,107 @@
+"""
+This script defines a PrompterTemplate class that assists in generating 
+conversation/prompt templates. The script facilitates formatting prompts 
+for inference and training by combining various context elements and user inputs.
+
+"""
+
+
+import dataclasses
+from typing import Dict, List, Tuple
+
+
+@dataclasses.dataclass
+class PrompterTemplate:
+    name: str
+    agent_behaviour: str = """
+    >>INTRODUCTION<<
+    {system_message}
+    """
+    roles: List[str] = ("USER", "ASSISTANT")
+
+    def __format_base(self, user_context: str, news_context: str) -> str:
+        """
+        Formats to Falcon compatible template, using `added_tokens`:
+        - INTRODUCTION : specifies agent behaviour
+        - COMMENT : used for context and user description
+        spec: https://huggingface.co/tiiuae/falcon-7b/raw/main/tokenizer.json
+        """
+        base_template = f"""
+        >>COMMENT<<
+        {user_context}
+        {news_context}
+        """
+        return base_template
+
+    def format_infer(self, sample: dict) -> Tuple[str, Dict[str, str]]:
+        """
+        - QUESTION: formats user question for the agent.
+        """
+        base_template = self.__format_base(
+            user_context=sample["about_me"], news_context=sample["context"]
+        )
+        question = f"""
+        >>QUESTION<<
+        {sample["question"]}
+        """
+
+        return {"prompt": base_template + question, "payload": sample}
+
+    def format_train(self, sample: dict) -> Tuple[str, Dict[str, str]]:
+        """
+        - QUESTION: formats user question for the agent.
+        - ANSWER: formats user format for the agent.
+        """
+        base_template = self.__format_base(
+            user_context=sample["about_me"], news_context=sample["context"]
+        )
+        question = f"""
+        >>QUESTION<<
+        {sample["question"]}
+        """
+
+        answer = f"""
+         >>ANSWER<<
+        {sample["response"]}
+        """
+
+        return {"prompt": base_template + question + answer, "payload": sample}
+
+    @property
+    def raw_template(self):
+        tmp = (
```

Comment:
Do we use this function anywhere?

Reply:
Yes, it's used in `inference.py` L:131 in order to log template blueprint to comet.ml LLM.
![Screenshot from 2023-09-06 19-01-48](https://github.com/iusztinpaul/hands-on-llms/assets/23241108/a880c090-c268-4fa1-93e8-e47868bee78a)

Reply:
I see. It is confusing that we have both this raw template and the template base function.

What about having a `train_raw_template` and `infer_raw_template` and dropping the `__format_base` method?

Thus, based on the scope, we can easily inject the variables (in the `format_train` and `format_infer` methods) by doing something like,

`train_raw_template.format(user_context, ...)`

For example, in the `inference.py` API, you use the `raw_template,` which is used for training when we need the inference template. 

Don't get me wrong, it is not a big bummer, but as we want to explain this code to others, I want to be as straightforward as possible.

Reply:
Solved, I've added a method `to_comet_llm()` that will return the template for LLM-Logging with comet-ml purposes.
The rest of the methods are descriptive, properties of PromptTemplate `train_raw_template` and `infer_raw_template`.

I've split the steps to creating a prompt into multiple chunks, so it would be easier to document & explain to future students.

Reply:
Looks good. But why can't we directly use `train_raw_template` or `infer_raw_template` when logging it to `comet`? The template class shouldn't be aware of Comet, and when logging it to comet, we should use the template used to generate the prompts, which are `train_raw_template` and `infer_raw_template`.

Except for this small detail, looks great :fire: